File: store/file_store_test.go
package store

import (
	"os"
	"path/filepath"
	"testing"
)

// createTempFileStore creates a new FileStore instance in a temporary directory
func createTempFileStore(t *testing.T) (Store, func()) {
	// Create a temporary directory for the test
	tempDir, err := os.MkdirTemp("", "vibekv-test-*")
	if err != nil {
		t.Fatalf("Failed to create temporary directory: %v", err)
	}

	// Create a file store in the temporary directory
	filePath := filepath.Join(tempDir, "vibekv.log")
	store, err := NewFileStore(filePath)
	if err != nil {
		os.RemoveAll(tempDir)
		t.Fatalf("Failed to create file store: %v", err)
	}

	// Return the store and a cleanup function
	cleanup := func() {
		store.Close()
		os.RemoveAll(tempDir)
	}

	return store, cleanup
}

// TestFileStorePut tests the Put operation for FileStore
func TestFileStorePut(t *testing.T) {
	store, cleanup := createTempFileStore(t)
	defer cleanup()
	TestPut(t, store)
}

// TestFileStoreGet tests the Get operation for FileStore
func TestFileStoreGet(t *testing.T) {
	store, cleanup := createTempFileStore(t)
	defer cleanup()
	TestGet(t, store)
}

// TestFileStoreDelete tests deleting an existing key for FileStore
func TestFileStoreDelete(t *testing.T) {
	store, cleanup := createTempFileStore(t)
	defer cleanup()
	TestDelete(t, store)
}

// TestFileStoreDeleteNonExistentKey tests deleting a non-existent key for FileStore
func TestFileStoreDeleteNonExistentKey(t *testing.T) {
	store, cleanup := createTempFileStore(t)
	defer cleanup()
	TestDeleteNonExistentKey(t, store)
}

// TestFileStoreRecovery tests that the store can recover its state from the transaction log
func TestFileStoreRecovery(t *testing.T) {
	// Create a temporary directory for the test
	tempDir, err := os.MkdirTemp("", "vibekv-recovery-test-*")
	if err != nil {
		t.Fatalf("Failed to create temporary directory: %v", err)
	}
	defer os.RemoveAll(tempDir)

	filePath := filepath.Join(tempDir, "vibekv.log")

	// Create a new file store
	store1, err := NewFileStore(filePath)
	if err != nil {
		t.Fatalf("Failed to create file store: %v", err)
	}

	// Add some data
	key := uint64(42)
	value := uint64(123)

	if err := store1.Put(key, value); err != nil {
		t.Fatalf("Failed to put key-value pair: %v", err)
	}

	// Close the store
	if err := store1.Close(); err != nil {
		t.Fatalf("Failed to close store: %v", err)
	}

	// Create a new store with the same file
	store2, err := NewFileStore(filePath)
	if err != nil {
		t.Fatalf("Failed to create file store: %v", err)
	}
	defer store2.Close()

	// Verify the data was recovered
	retrievedValue, err := store2.Get(key)
	if err != nil {
		t.Fatalf("Failed to get value for key: %v", err)
	}

	if retrievedValue != value {
		t.Errorf("Retrieved value %d does not match original value %d", retrievedValue, value)
	}
}

// TestFileStoreComplexRecovery tests recovery with multiple operations including deletions
func TestFileStoreComplexRecovery(t *testing.T) {
	// Create a temporary directory for the test
	tempDir, err := os.MkdirTemp("", "vibekv-complex-recovery-test-*")
	if err != nil {
		t.Fatalf("Failed to create temporary directory: %v", err)
	}
	defer os.RemoveAll(tempDir)

	filePath := filepath.Join(tempDir, "vibekv.log")

	// Create a new file store
	store1, err := NewFileStore(filePath)
	if err != nil {
		t.Fatalf("Failed to create file store: %v", err)
	}

	// Add multiple key-value pairs
	keys := []uint64{1, 2, 3, 4, 5}
	values := []uint64{10, 20, 30, 40, 50}

	for i, key := range keys {
		if err := store1.Put(key, values[i]); err != nil {
			t.Fatalf("Failed to put key-value pair: %v", err)
		}
	}

	// Update a key
	if err := store1.Put(keys[1], 25); err != nil {
		t.Fatalf("Failed to update key: %v", err)
	}

	// Delete a key
	if err := store1.Delete(keys[2]); err != nil {
		t.Fatalf("Failed to delete key: %v", err)
	}

	// Close the store
	if err := store1.Close(); err != nil {
		t.Fatalf("Failed to close store: %v", err)
	}

	// Create a new store with the same file
	store2, err := NewFileStore(filePath)
	if err != nil {
		t.Fatalf("Failed to create file store: %v", err)
	}
	defer store2.Close()

	// Verify the data was recovered correctly
	// Check the updated key
	value, err := store2.Get(keys[1])
	if err != nil {
		t.Fatalf("Failed to get updated key: %v", err)
	}
	if value != 25 {
		t.Errorf("Retrieved value %d for updated key does not match expected value 25", value)
	}

	// Check the deleted key
	_, err = store2.Get(keys[2])
	if err == nil {
		t.Errorf("Expected error when getting deleted key, but got nil")
	}

	// Check other keys
	expectedValues := map[uint64]uint64{
		1: 10,
		4: 40,
		5: 50,
	}

	for k, expectedValue := range expectedValues {
		value, err := store2.Get(k)
		if err != nil {
			t.Fatalf("Failed to get key %d: %v", k, err)
		}
		if value != expectedValue {
			t.Errorf("Retrieved value %d for key %d does not match expected value %d", value, k, expectedValue)
		}
	}
}

File: haskell/test/VibeKV/FileStoreTest.hs
module VibeKV.FileStoreTest
  ( tests
  ) where

import Test.Tasty
import Test.Tasty.HUnit
import VibeKV.Store
import VibeKV.FileStore
import qualified VibeKV.StoreTest as StoreTest
import System.IO.Temp
import System.FilePath
import System.Directory

tests :: TestTree
tests = testGroup "FileStore Tests"
  [ testPut
  , testGet
  , testDelete
  , testDeleteNonExistentKey
  , testRecovery
  , testComplexRecovery
  ]

-- | Test the Put operation for FileStore
testPut :: TestTree
testPut = withTempFileStore $ \store cleanup ->
  StoreTest.testPut "FileStore" store cleanup

-- | Test the Get operation for FileStore
testGet :: TestTree
testGet = withTempFileStore $ \store cleanup ->
  StoreTest.testGet "FileStore" store cleanup

-- | Test the Delete operation for FileStore
testDelete :: TestTree
testDelete = withTempFileStore $ \store cleanup ->
  StoreTest.testDelete "FileStore" store cleanup

-- | Test deleting a non-existent key for FileStore
testDeleteNonExistentKey :: TestTree
testDeleteNonExistentKey = withTempFileStore $ \store cleanup ->
  StoreTest.testDeleteNonExistentKey "FileStore" store cleanup

-- | Test recovery from the transaction log
testRecovery :: TestTree
testRecovery = testCase "FileStore - Recovery" $ do
  withSystemTempDirectory "vibekv-test" $ \tempDir -> do
    let filePath = tempDir </> "vibekv.log"
    
    -- Create a new file store
    store1 <- newFileStore filePath
    
    -- Add some data
    let key = 42
        value = 123
    
    store1' <- put store1 key value
    
    -- Close the store
    close store1'
    
    -- Create a new store with the same file
    store2 <- newFileStore filePath
    
    -- Verify the data was recovered
    result <- get store2 key
    case result of
      Right retrievedValue -> 
        assertEqual "Retrieved value should match original" value retrievedValue
      Left KeyNotFound -> 
        assertFailure "Key should exist after recovery"
    
    -- Close the store
    close store2

-- | Test recovery with multiple operations including deletions
testComplexRecovery :: TestTree
testComplexRecovery = testCase "FileStore - Complex Recovery" $ do
  withSystemTempDirectory "vibekv-test" $ \tempDir -> do
    let filePath = tempDir </> "vibekv.log"
    
    -- Create a new file store
    store1 <- newFileStore filePath
    
    -- Add multiple key-value pairs
    let keys = [1, 2, 3, 4, 5]
        values = [10, 20, 30, 40, 50]
    
    store1' <- foldM (\s (k, v) -> put s k v) store1 (zip keys values)
    
    -- Update a key
    store1'' <- put store1' 2 25
    
    -- Delete a key
    Right store1''' <- delete store1'' 3
    
    -- Close the store
    close store1'''
    
    -- Create a new store with the same file
    store2 <- newFileStore filePath
    
    -- Verify the data was recovered correctly
    -- Check the updated key
    result1 <- get store2 2
    case result1 of
      Right value -> 
        assertEqual "Updated value should be 25" 25 value
      Left KeyNotFound -> 
        assertFailure "Updated key should exist after recovery"
    
    -- Check the deleted key
    result2 <- get store2 3
    case result2 of
      Right _ -> 
        assertFailure "Deleted key should not exist after recovery"
      Left KeyNotFound -> 
        return ()
    
    -- Check other keys
    let expectedValues = [(1, 10), (4, 40), (5, 50)]
    
    forM_ expectedValues $ \(k, expectedValue) -> do
      result <- get store2 k
      case result of
        Right value -> 
          assertEqual ("Value for key " ++ show k) expectedValue value
        Left KeyNotFound -> 
          assertFailure ("Key " ++ show k ++ " should exist after recovery")
    
    -- Close the store
    close store2

-- | Helper function to create a FileStore in a temporary directory for testing
withTempFileStore :: (FileStore -> (FileStore -> IO ()) -> TestTree) -> TestTree
withTempFileStore f = do
  let setup = do
        tempDir <- getCanonicalTemporaryDirectory >>= flip createTempDirectory "vibekv-test"
        let filePath = tempDir </> "vibekv.log"
        store <- newFileStore filePath
        let cleanup s = do
              close s
              removeDirectoryRecursive tempDir
        return (store, cleanup)
  
  withResource setup (\_ -> return ()) $ \getResource ->
    testCase "FileStore" $ do
      (store, cleanup) <- getResource
      let test = f store cleanup
      return test

-- Utility functions
foldM :: Monad m => (b -> a -> m b) -> b -> [a] -> m b
foldM _ z [] = return z
foldM f z (x:xs) = do
  z' <- f z x
  foldM f z' xs

forM_ :: Monad m => [a] -> (a -> m b) -> m ()
forM_ xs f = mapM_ f xs
